flow/diff_context.cc

```diff
@@ -45,6 +45,10 @@ void DiffContext::PushTransform(const SkMatrix& transform) {
   }
 }
 
+void DiffContext::SetTransform(const SkMatrix& transform) {
+  state_.transform = transform;
```

Comment:
While this is currently only used for adjusting the integer translation components, the cull_rect was maintained using the non-integer transform. Is that a problem?

Reply:
This is a very good question. I don't think so

1. For culling during painting in Layer `needs_painting` we check layer bounds against canvas clip rect. None of these are affected by `SUPPORT_FRACTIONAL_TRANSLATION`. 

2. In `DiffContext::AddLayerBounds`, we check if `rect` (layer bounds) intersects `cull_rect`, but again, none of these are affected by `SUPPORT_FRACTIONAL_TRANSLATION`. Only after we determine that they intersect, we transform the bounds to screen coordinates, using the adjusted transform.

This should also work for `OpacityLayer`. That is because `DiffContext::PushTransform` affects the cull rect (which should be independent of `SUPPORT_FRACTIONAL_TRANSLATION` just like it is when culling layers in `needs_painting`), while `DiffContext::SetTransform` only affects calculation of painted area (the transform matrix is only used for transforming layer bounds to screen coordinates).

Reply:
Actually, I might be wrong about the `OpacityLayer`, since the layer sets the leaf transform to integer, which I assume does affect `needs_painting` on child layers.

Reply:
My point is mainly that PushTransform feels the needs to update the cull rect based on the new transform. SetTransform does not do that.

It may not matter for our current uses, but that is definitely a bug of some sort.

In general I don't like to untransform cull_rects and bounds if it is at all possible, preferring to go the other way to minimize multiple-transform bounding box inflation.

Reply:
I also note that the bounds for the layer are not adjusted for fractional transforms. This is particularly difficult because it isn't just recording them with a bias like we do for the offset - the test for needs_painting() would likely need to be aware of the manipulation.

This is an issue outside the scope of DiffContext, but these changes brought this to my attention.

Reply:
Current behavior with `DiffContext is consistent with what `PictureLayer` does for example. Culling (in `needs_painting`) is done with original bounds (non integer CTM), but after then layer is painted slightly off (integer CTM).

This is same behavior you get with `DiffContext` currently. But it seems more like by accident than by design. Also the more I think about it the less I am convincend that it does the right thing with `OpacityLayer`.

But there is also another `cull_rect` used during preroll (mostly used for checking raster cache IIRC) and that one is also not adjusted for `SUPPORT_FRACTIONAL_TRANSLATION`, might be yet another subtle issue.

Reply:
I filed https://github.com/flutter/flutter/issues/91260 on the preroll bounds vs FRACTIONAL_TRANSLATION issue.

Reply:
If PictureLayer and DisplayListLayer compute correct bounds in Preroll (adjusted for integer CTM) and OpacityLayer does same thing on child_matrix in Preroll (snap translation to integer), then as far as I can tell DiffContext doesn't need to know anything about SUPPORT_FRACTIONAL_TRANSLATION at all.

Reply:
Doesn't DiffContext need to know which bounds need repainting? It has the same issue. If it is off by 1 then we leave dirt behind when an item disappears, or we shave a pixel off when it appears (and similarly we leave a trail behind when it is moving).

The thing is that the paint_bounds() are supposed to be relative to the transform that came to the layer from above. This is why the bounds are moved by the offset. The paint() call will apply the `offset` property, so the bounds (which are subject to the offset when painting) need to be translated by that offset. Similarly the paint() call will also apply the sub-pixel rounding - thereby making the painted bounds really subject to `offset`+fract_offset. The reported bounds computed in preroll(), though, only adjust the bounds of the child by the `offset` property, but leave out the fractional offset that will be implicitly applied by the ifdef.

This all matters because `needs_painting` will only take into account the matrix that comes to the layer (before the layer gets a chance to modify it) + the reported bounds. If the reported bounds are off by a fraction of a pixel, then it isn't doing the culling check properly.

Reply:
Basically we paint at:
  layer_transform -> offset -> fractional_adjustment -> child
needs_painting tests
  layer_transform -> paint_bounds    <vs>    cull_rect
For the paint_bounds to properly represent the area being painted, they should be:
  offset -> fractional_adjustment -> child's bounds
But, instead they are computed by preroll as:
  offset -> child's bounds

Both paint_bounds and the data that DiffContext collects should track:
  offset -> fractional_adjustment -> child's bounds
preroll does not do that, but these mods to DiffContext hopefully do that.

Reply:
To be precise. Right now DiffContext doesn't use `paint_bounds`, for picture and DL layer it uses `PushTransform` for picture translation and then `picture()->cullRect()` for bounds. That's in part because diffing happens before preroll, but I think the order can be easily reversed.

Now if we compute layer `paint_bounds` properly, so that when parent CTM is applied it covers the actual painted area (including integer translation), DiffContext could just be using these `paint_bounds` instead of transform and picture cull rect and it should get correct result.

So the question is, can we compute layer `paint_bounds` in preroll in a way where it actually reflects area where the layer will paint? i.e. take the `paint_bounds` we currently have, apply `RasterCache::GetIntegralTransCTM(matrix)` and then apply inverse of `matrix`. This should get us correct layer `paint_bounds` (and should also fix possible problem with `needs_paint()` which would now check the clip rect against correct paint bounds.

Reply:
Yes, preroll can do that. That's why I submitted the other issue. I'm not sure we can get to it very soon with everything else and the idea that's floating about to get rid of the FRACTIONAL ifdef code.

To be very very clear, though. double transforming bounds is a really bad idea. Every transform of bounds that isn't a scale/translate greatly inflates the result. transform/inverse_transform doubles down on that inflation of the bounds.

In any case, even if the transform isn't a pure scale/translate the end result you are looking for is a simple translation of the bounds and so a more direct route for calculating that offset should be possible.

Reply:
To assess where this PR is at, I think with these changes, you are setting up the environment to correctly compute the true rendered bounds in the DiffContext, so this part is fixed. The rest of this discussion is how to fix paint_bounds() which isn't a concern for this PR.

That still leaves the validity of the cull_rect. It won't have been adjusted for the rounding of the translation components. If it is only used for optimization, transform/untransforming won't break anything, especially in a scale/translate situation, but a more accurate adjustment would be to do the math I talked about above which is to take the device distance that the translation components were adjusted by and use a inverse delta transform on them.

Do we still do Diff-ing if there is a perspective transform at play?

Reply:
If there is a chance for properly fixing `paint_bounds` that would make this PR obsolete. 

This PR will properly calculate paint bounds of Picture/DL layers, but it will not properly cull paint bounds of child layers inside `OpacityLayer`. I'll have to think that one through.

Reply:
> Do we still do Diff-ing if there is a perspective transform at play?

DiffContext doesn't really care about that. Perspective transforms are not an issue during diffing.

Reply:
> If there is a chance for properly fixing `paint_bounds` that would make this PR obsolete.

How is that? You mentioned earlier that Diff is done before preroll, so minimally something would need to be done, no>

> This PR will properly calculate paint bounds of Picture/DL layers, but it will not properly cull paint bounds of child layers inside `OpacityLayer`. I'll have to think that one through.

> DiffContext doesn't really care about that. Perspective transforms are not an issue during diffing.

In the case of perspective transforms, then, the trick of inverse transforming the rounding of the CTM would not work.

A technique that would work with minimal expansion of bounds in all cases would be to turn the bounds into a quad and then transform and untransform the points of the quad before turning it back into a rect. Arguably, that would be better for cases where a given cull rect is transformed through multiple transform_layers. Perhaps we should be keeping a cull_quad instead of a cull_rect and then only make a rect out of it when we need to test against it or intersect it?

Reply:
> > If there is a chance for properly fixing `paint_bounds` that would make this PR obsolete.
> 
> How is that? You mentioned earlier that Diff is done before preroll, so minimally something would need to be done, no>

You're right. They are independent steps in current PR, which confused me for a bit, but that's wrong. I forgot to set the computed clip rect as cull rect for preroll context. Perhaps it got lost after those numerous rebases :) The cull rect is necessary to not rasterize layers (`TryToPrepareRasterCache`) that are outside of what we're painting.

So unless we can get rid of `SUPPORT_FRACTIONAL_TRANSLATION` completely this needs to be addressed here.

> 
> > This PR will properly calculate paint bounds of Picture/DL layers, but it will not properly cull paint bounds of child layers inside `OpacityLayer`. I'll have to think that one through.
> 
> > DiffContext doesn't really care about that. Perspective transforms are not an issue during diffing.
> 
> In the case of perspective transforms, then, the trick of inverse transforming the rounding of the CTM would not work.

I don't quite understand why we would need to round perspective CTM. `RasterCache::GetIntegralTransCTM` only modifies the matrix if `ctm.isScaleTranslate()`. 
 
> A technique that would work with minimal expansion of bounds in all cases would be to turn the bounds into a quad and then transform and untransform the points of the quad before turning it back into a rect. Arguably, that would be better for cases where a given cull rect is transformed through multiple transform_layers. Perhaps we should be keeping a cull_quad instead of a cull_rect and then only make a rect out of it when we need to test against it or intersect it?

I'm not sure what you mean by expansion of bounds here. During diffing the cull rect is always kept in current layer coordinate space. So for culling we always check intersection of layer bounds vs cull rect (in layer space, not sceen space).

Reply:
Btw. For culling during paint, `Layer::needs_painting` calls `context.leaf_nodes_canvas->quickReject(paint_bounds_)`. `quickReject` takes paint bounds, transform them to device coordinates and checks intersection with `fQuickRejectBounds`, which is kept in device coordinates. We could very easily do culling during diffing same way. This way we wouldn't need to adjust cull rect at all. The `DiffContext` code where cull rect is kept in layer coordinates was written before @flar's change to `needs_painting`.

Reply:
I pushed two commits.

[First one](https://github.com/flutter/engine/pull/29001/commits/c8b0a7ce688be12c448034f60b9d04de75949bfd) changes `cull_rect` to device coordinates. This has two benefits:

1. It is consistent with how `Layer::needs_painting` works
2. It means that `cull_rect` no longer depends on current transform and does not need to update every time transform changes.

[Second one](https://github.com/flutter/engine/pull/29001/commits/3b4f03b1a8a2b76cd5f0aa0cfa6401f9614bfa13) addresses the fact that during painting we first cull layer (`Layer::needs_painting`) while using the original transform, but afterwards we paint it with Integer CTM. So diff context stores both the original transform and the overridden one, uses original for culling and overridden for paint coordinates.

This should replicate the paint process closely, I'll test this some more tomorrow.